test/end2end_test.go

```diff
@@ -4636,6 +4636,80 @@ func testClientRequestBodyErrorCancelStreamingInput(t *testing.T, e env) {
 	})
 }
 
+func (s) TestClientInitialHeaderEndStream(t *testing.T) {
+	for _, e := range listTestEnv() {
+		if e.httpHandler {
+			continue
+		}
+		testClientInitialHeaderEndStream(t, e)
+	}
+}
+
+func testClientInitialHeaderEndStream(t *testing.T, e env) {
+	te := newTest(t, e)
+	ts := &funcServer{streamingInputCall: func(stream testpb.TestService_StreamingInputCallServer) error {
+		_, err := stream.Recv()
+		if err == nil {
+			errUnexpectedData := errors.New("unexpected data received in func server method")
+			t.Error(errUnexpectedData)
+			return errUnexpectedData
+		}
+		return nil
+	}}
+	te.startServer(ts)
+	defer te.tearDown()
+	te.withServerTester(func(st *serverTester) {
+		// Send a headers with END_STREAM flag, but then write data.
+		st.writeHeadersGRPC(1, "/grpc.testing.TestService/StreamingInputCall", true)
+		st.writeData(1, false, []byte{0, 0, 0, 0, 0})
+		st.wantAnyFrame()
+		st.wantAnyFrame()
+		st.wantRSTStream(http2.ErrCodeStreamClosed)
```

Comment:
This looks potentially fragile, or like it might not be testing the intended behavior sometimes (maybe always?).  The handler should not exit as soon as `Recv()` returns `nil`, or else it could be sending this RST_STREAM only because the server terminated the stream entirely.

Looks like this is the same as what you changed in the test below?

Reply:
I added some synchronization here, so it will always check for the RST before handler returns.